### PR TITLE
Support runnables in langchain

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/10521?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10521/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10521
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Support RunnableSequence and other reasonable runnables within langchain flavor for save/load.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -325,79 +256,6 @@ def load_retriever(persist_directory):
     _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
 
 
-def _validate_and_wrap_lc_model(lc_model, loader_fn):
```
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:328]: This part is moved to utils.py without change

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -107,53 +84,6 @@ def get_default_conda_env():
...
-
-def _get_special_chain_info_or_none(chain):
-    for special_chain_class, loader_arg in _get_map_of_special_chain_class_to_loader_arg().items():
-        if isinstance(chain, special_chain_class):
-            return _SpecialChainInfo(loader_arg=loader_arg)
-
-
-@functools.lru_cache
-def _get_map_of_special_chain_class_to_loader_arg():
-    import langchain
```
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:122]: This part is moved to utils.py without change

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -555,136 +413,27 @@ def load_retriever(persist_directory):
 
 
 def _save_model(model, path, loader_fn, persist_dir):
-    import langchain
-
```
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:559]: This part is moved into utils.py _save_base_lcs function without change.

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -555,136 +413,27 @@ def load_retriever(persist_directory):
...
-    path,
-    model_type,
-    loader_arg=None,
-    agent_path=None,
-    tools_path=None,
-    agent_primitive_path=None,
-    loader_fn_path=None,
-    persist_dir=None,
-):
-    from langchain.chains.loading import load_chain
```
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:650]: Lines inside this function are moved into _load_base_lcs with no change.

File: mlflow/langchain/runnables.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        else:
+            steps_conf[step][_MODEL_TYPE_KEY] = runnable.__class__.__name__
+            # Save some simple runnables that langchain natively supports.
+            if hasattr(runnable, "save"):
+                runnable.save(f"{save_path!s}/{step}.yaml")
+            # TODO: check if `dict` is enough to load it back
+            elif hasattr(runnable, "dict"):
+                runnable_dict = runnable.dict()
+                with open(f"{save_path!s}/{step}.yaml", "w") as f:
+                    yaml.dump(runnable_dict, f, default_flow_style=False)
```
Comment (daniellok-db) [inline mlflow/langchain/runnables.py:173]:
```diff
            steps_conf[step][_MODEL_TYPE_KEY] = runnable.__class__.__name__
+           save_lc_native_path = f"{save_path!s}/{step}.yaml"
            # Save some simple runnables that langchain natively supports.
            if hasattr(runnable, "save"):
-               runnable.save(f"{save_path!s}/{step}.yaml")
+               runnable.save(save_lc_native_path)
            # TODO: check if `dict` is enough to load it back
            elif hasattr(runnable, "dict"):
                runnable_dict = runnable.dict()
-               with open(f"{save_path!s}/{step}.yaml", "w") as f:
+               with open(save_lc_native_path, "w") as f:
                    yaml.dump(runnable_dict, f, default_flow_style=False)
```

Comment (daniellok-db) [reply mlflow/langchain/runnables.py:173]: we could maybe const this path

File: mlflow/langchain/runnables.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            else:
+                unsaved_runnables[step] = f"{runnable!r}"
```
Comment (harupy) [inline mlflow/langchain/runnables.py:175]: Can we throw immediately if we find an unserializable runnable?

File: mlflow/langchain/runnables.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+    if unsaved_runnables:
+        raise MlflowException(
+            f"Failed to save runnable sequence: {unsaved_runnables}. "
+            "Runnable must have either `save` or `dict` method."
+        )
+
+    with open(f"{save_path!s}/{_RUNNABLE_SEQUENCE_STEPS_FILE_NAME}", "w") as f:
```
Comment (harupy) [inline mlflow/langchain/runnables.py:183]:
```suggestion
    with save_path.joinpath(_RUNNABLE_SEQUENCE_STEPS_FILE_NAME).open("w") as f:
```

File: mlflow/langchain/runnables.py (hunk 6 above)
Comment (daniellok-db) [reply mlflow/langchain/runnables.py:175]: i guess the benefit of not doing that is that the user can see all the types of unsupported runnables

File: mlflow/langchain/runnables.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            f"File {steps_conf_file} must exist in order to load the RunnableSequence."
+        )
+    steps_conf = _load_from_yaml(steps_conf_file)
+    if model_type == RunnableSequence.__name__:
+        steps = []
+    elif model_type == RunnableParallel.__name__:
+        steps = {}
+    else:
+        raise MlflowException(f"Unsupported model type {model_type}. ")
+    for file in sorted(os.listdir(load_path)):
```
Comment (daniellok-db) [inline mlflow/langchain/runnables.py:99]: if the runnable has more than 10 steps, will this result in weird sorting? e.g`["0", "1", "10", "2", "3" ...]`

File: mlflow/langchain/runnables.py (hunk 6 above)
Comment (harupy) [reply mlflow/langchain/runnables.py:175]: makes sense

Comment (harupy) [inline mlflow/langchain/runnables.py:175]:
```suggestion
                unsaved_runnables[step] = repr(runnable)
```

Comment (harupy) [reply mlflow/langchain/runnables.py:175]: or

```suggestion
                unsaved_runnables[step] = str(runnable)
```

File: mlflow/langchain/runnables.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def _save_pickable_runnable(model, path):
+    if not path.endswith(".pkl"):
+        raise ValueError(f"File path must end with .pkl, got {path!s}.")
+    with open(path, "wb") as f:
+        cloudpickle.dump(model, f)
+
+
+def _save_runnables(model, path, model_data_path=None, loader_fn=None, persist_dir=None):
```
Comment (daniellok-db) [inline mlflow/langchain/runnables.py:194]: not sure how others feel, but i'd find it helpful if we could include some docstring outlining the layout of the saved YAML file / directory structure. i don't think it's a must-have, but it did take some reading for me to understand how we're structuring it

**Comment by daniellok-db** [comment]
Left a few comments but generally the logic looks good to me, awesome work! Let's get confirmation on the sort order for > 10 steps, and then we can probably ship it

File: mlflow/langchain/utils.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+
+
+@lru_cache
+def custom_type_to_loader_dict():
+    # helper function to load output_parsers from config
+    def _load_output_parser(config: dict) -> dict:
+        """Load output parser."""
+        from langchain.schema.output_parser import StrOutputParser
+
+        output_parser_type = config.pop("_type")
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/langchain/utils.py:111]: Yes in most cases, otherwise it can't be saved

File: mlflow/langchain/runnables.py (hunk 9 above)
Comment (serena-ruan) [reply mlflow/langchain/runnables.py:194]: Yea definitely! I'll add some docstrings to explain this

File: mlflow/langchain/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/langchain/utils.py:111]: Can we use `config.pop("_type", None)`?

File: mlflow/langchain/runnables.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    :param steps: List of steps to construct the RunnableSequence from.
+    """
+    from langchain.schema.runnable import RunnableSequence
+
+    if len(steps) < 2:
+        raise ValueError(f"RunnableSequence must have at least 2 steps, got {len(steps)}.")
+
+    if len(steps) == 2:
+        return RunnableSequence(first=steps[0], last=steps[1])
+    return RunnableSequence(first=steps[0], middle=steps[1:-1], last=steps[-1])
```
Comment (harupy) [inline mlflow/langchain/runnables.py:128]:
```suggestion
    first, *middle, last = steps
    return RunnableSequence(first=first, middle=middle, last=last)
```

File: .github/workflows/cross-version-tests.yml [hunk 12]
```diff
@@ -187,6 +187,12 @@ jobs:
       - name: Install ${{ matrix.package }} ${{ matrix.version }}
         run: |
           ${{ matrix.install }}
+      - name: Install pydantic no-binary for langchain
+        if: matrix.package == 'langchain' && ( matrix.version == 'dev' || matrix.version >= '0.0.267')
+        run: |
+          pip uninstall pydantic -y
+          # 0.0.267 is the minimum version that supports Pydantic v2
+          pip install pydantic --no-binary :all:
```
Comment (harupy) [inline .github/workflows/cross-version-tests.yml:195]: can we add a comment explaining why no-binary is requierd?

File: mlflow/langchain/runnables.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    for key, runnable in generator:
+        step = str(key)
+        steps_conf[step] = {}
+        if isinstance(runnable, lc_runnables_types()):
+            steps_conf[step][_MODEL_TYPE_KEY] = runnable.__class__.__name__
+            steps_conf[step].update(
+                _save_runnables(runnable, save_path, step, loader_fn, persist_dir)
+            )
+        elif isinstance(runnable, base_lc_types()):
+            save_runnable_path = f"{save_path!s}/{step}"
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/langchain/runnables.py:176]: Can we create a directory named `steps` and save the steps there?

File: mlflow/langchain/runnables.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+def _save_runnable_with_steps(steps, file_path: Union[Path, str], loader_fn=None, persist_dir=None):
+    """
+    Save the model with steps. Currently it supports saving RunnableSequence and RunnableParallel.
+    If saving a RunnableSequence, steps is a list of Runnable objects. We save each step to the
+    path (either a folder or a single file) named by the step index.
+    e.g.
+        model - model/0.yaml
+              - model/1.pkl
+              - model/2 -- model/2/model.yaml
+              - steps.yaml
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/langchain/runnables.py:144]: For RunnableParallel, it's key-runnable pairs, do you want to keep current behavior or also save them into seperate folders?
- steps
  - context
    - model.yml
  - question
    - model.pkl

File: mlflow/langchain/runnables.py [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+        )
+    steps_conf = _load_from_yaml(steps_conf_file)
+    steps_path = load_path / _STEPS_FOLDER_NAME
+    if not steps_path.exists() or not steps_path.is_dir():
+        raise MlflowException(
+            f"Folder {steps_path} must exist and must be a directory "
+            "in order to load runnable with steps."
+        )
+    if model_type == RunnableSequence.__name__:
+        steps = []
```
Comment (harupy) [inline mlflow/langchain/runnables.py:112]: Is it possible to always use a dict? For `RunnableSequence`, we can sort values by key.

File: tests/langchain/test_langchain_model_export.py [hunk 16]
```diff
@@ -768,3 +771,412 @@ def test_agent_with_unpicklable_tools(tmp_path):
...
+    from langchain.prompts import ChatPromptTemplate
+    from langchain.schema.output_parser import StrOutputParser
+
+    from mlflow.langchain.utils import _fake_simple_chat_model
+
+    prompt = ChatPromptTemplate.from_template(
+        "What is a good name for a company that makes {product}?"
+    )
+    chat_model = _fake_simple_chat_model()()
+    chain = prompt | chat_model | StrOutputParser()
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:1076]: Is this the example that needs to work for the demo?

Comment (serena-ruan) [reply tests/langchain/test_langchain_model_export.py:1076]: test_save_load_rag is a similar one

File: mlflow/langchain/runnables.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+
+    config_path = os.path.join(path, model_config.get(_MODEL_DATA_KEY, _MODEL_DATA_YAML_FILE_NAME))
+    # Load runnables from config file
+    if config_path.endswith(".yaml"):
+        config = _load_from_yaml(config_path)
+    elif config_path.endswith(".json"):
+        config = _load_from_json(config_path)
+    else:
+        raise MlflowException(
+            f"Cannot load runnable without a config file. Got path {config_path!s}."
```
Comment (harupy) [inline mlflow/langchain/runnables.py:56]:
```suggestion
            f"Cannot load runnable without a config file. Got path {config_path}."
```

File: mlflow/pyfunc/__init__.py [hunk 18]
```diff
@@ -1506,6 +1510,8 @@ def udf(
             os.environ.update(openai_env_vars)
         if mlflow_testing:
             _MLFLOW_TESTING.set(mlflow_testing)
+        if env_vars:
+            os.environ.update(env_vars)
```
Comment (WeichenXu123) [inline mlflow/pyfunc/__init__.py:1514]: pyspark UDF worker might be reused by other spark jobs, see `spark.python.worker.reuse` config (default True), so we need to ensure the env is restored once UDF completes.

File: mlflow/pyfunc/__init__.py [hunk 19]
```diff
@@ -1108,6 +1108,7 @@ def spark_udf(
     result_type=None,
     env_manager=_EnvManager.LOCAL,
     params: Optional[Dict[str, Any]] = None,
+    env_vars=None,
```
Comment (WeichenXu123) [inline mlflow/pyfunc/__init__.py:1111]: The argument name is not very clear, what about renaming it to `extra_udf_envs` ?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py]: I'll raise a separate PR about this change as it's irrelevant to langchain

**Pull request merged**